packages/crypto/src/bls12_318/aggregate.rs

```diff
@@ -0,0 +1,229 @@
+        .chunks_exact(G2_POINT_SIZE)
+        .map(|data| {
+            let mut buf = [0u8; 96];
+            buf[..].clone_from_slice(data);
+            buf
+        })
+        .collect();
+
+    let mut decoded_points = Vec::with_capacity(points_count);
+    points
+        .par_iter()
+        .map(g2_from_fixed)
+        .collect_into_vec(&mut decoded_points);
+
+    let out: Result<Vec<G2>, CryptoError> = decoded_points.into_iter().collect();
+    let out = out?;
+
+    let out = g2_sum(&out);
+
+    Ok(out.to_compressed())
+}
+
+/// Creates a sum of points in G1.
+///
+/// This is fast since math is done on projective points. Parallelization does not help here
+/// for ~500 elements.
+#[inline]
+pub fn g1_sum(elements: &[G1]) -> G1 {
+    elements.iter().sum()
+}
+
+/// Creates a sum of points in G2.
+///
+/// This is fast since math is done on projective points. Parallelization does not help here
+/// for ~500 elements.
+#[inline]
+pub fn g2_sum(elements: &[G2]) -> G2 {
+    elements.iter().sum()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::points::{g1_from_variable, g1s_from_variable};
+    use super::*;
+    use base64::engine::general_purpose::STANDARD;
+    use base64_serde::base64_serde_type;
+    use hex_literal::hex;
+
+    base64_serde_type!(Base64Standard, STANDARD);
+
+    #[derive(Debug, PartialEq, serde::Serialize, serde::Deserialize)]
+    struct EthPubkey(#[serde(with = "Base64Standard")] Vec<u8>);
+
+    #[derive(Debug, PartialEq, serde::Serialize, serde::Deserialize)]
+    struct EthHeaders {
+        public_keys: Vec<EthPubkey>,
+        #[serde(with = "Base64Standard")]
+        message: Vec<u8>,
+        #[serde(with = "Base64Standard")]
+        signature: Vec<u8>,
+        #[serde(with = "Base64Standard")]
+        aggregate_pubkey: Vec<u8>,
+    }
+
+    const ETH_HEADER_FILE: &str =
+        include_str!("../../testdata/eth-headers/1699693797.394876721s.json");
+
+    fn read_eth_header_file() -> EthHeaders {
+        serde_json::from_str(ETH_HEADER_FILE).unwrap()
+    }
+
+    /// Arbitrary point in G1
+    fn p1() -> G1 {
+        // Public key of classic League of Entropy Mainnet (curl -sS https://drand.cloudflare.com/info)
+        g1_from_fixed(&hex!("868f005eb8e6e4ca0a47c8a77ceaa5309a47978a7c71bc5cce96366b5d7a569937c529eeda66c7293784a9402801af31")).unwrap()
+    }
+
+    /// Arbitrary point in G2
+    fn p2() -> G2 {
+        g2_from_fixed(&hex!("b6ed936746e01f8ecf281f020953fbf1f01debd5657c4a383940b020b26507f6076334f91e2366c96e9ab279fb5158090352ea1c5b0c9274504f4f0e7053af24802e51e4568d164fe986834f41e55c8e850ce1f98458c0cfc9ab380b55285a55")).unwrap()
+    }
+
+    #[test]
+    fn bls12_318_aggregate_g1_works() {
+        let sum = bls12_381_aggregate_g1(b"").unwrap();
+        assert_eq!(sum, G1::identity().to_compressed());
+    }
+
+    #[test]
+    fn bls12_318_aggregate_g2_works() {
+        let sum = bls12_381_aggregate_g2(b"").unwrap();
+        assert_eq!(sum, G2::identity().to_compressed());
```

Comment:
Let's turn the empty input into an error case for `bls12_381_aggregate_g1`/`bls12_381_aggregate_g2`. This is safer for our contract developers where an empty list is unlikely to be intended.

Also for those two tests some unit tests with elements would be great.

Reply:
They are tested inside the integration test file `bls12_381.rs` against test vectors from eth2book.info, that should be sufficient, no?